_posts/2025-04-25-from-recipes-to-vectors-building-a-hybrid-search-app-with-opensearch.md

```diff
@@ -0,0 +1,299 @@
+
+## Follow along
+
+All the code in this article is available on [GitHub](https://github.com/BigDataBoutique/opensearch-semantic-search-tutorial) for reference.
+To follow along, you'll need:
+* A kaggle account and API token
+* A Cohere account and API token
+* An OpenSearch 2.17 cluster. You can easily set it up on [AWS](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/gsg.html).
+
+## Choosing a Dataset
+
+Your dataset might already be stored in OpenSearch, be in another database, or you might not have created it yet. For this article, I’ll use a publicly available **recipe** dataset from Kaggle.
+We'll be working with the following columns:
+* *Name* – A key identifier for search, but also interesting for semantic search
+* *DatePublished* – Useful for filtering and sorting
+* *AuthorName* – Another potential search and filter field
+* *TotalTime* - Useful for filtering out long or short recipes
+* *Ingredients* – Great for keyword-based search but we also want to search for ingredients semantically.
+* *RecipeCategory* – Useful for categorization and filtering
+
+### Why This Dataset?
+
+I chose this dataset because it has both structured and unstructured data, making it a good example for keyword and semantic search. Fields like publish date, ingredients, and category work well for traditional filtering and keyword-based search, while the instructions provide long text, which is useful for testing semantic search.
+Another reason I picked this dataset is that it is already clean, so we don’t need much preprocessing. In real-world projects, your data may not be this clean and could need extra work before you can use it. How much cleaning is needed depends on your specific data and use case, so keep that in mind when preparing your dataset.
+
+## Choosing an embedding model
+
+Once we have our dataset, the next step is to choose an embedding model to turn our text into vectors. This choice affects search quality and system performance, so it’s important to think about pricing, supported languages, input size limits, and output vector size. These things depend on how complex and long your text is. For example, models with larger input limits are better for long documents, while higher-dimensional vectors can improve search accuracy but need more storage and processing power.
+
+Sometimes, you may need to fine-tune an embedding model if the available ones do not understand your specific domain or terminology well. This is useful for technical, medical, or legal texts, where general models may not capture important details. However, fine-tuning requires a lot of labeled data and extra training costs, so it is best when standard models do not give good results.
+
+For this tutorial, I chose to use Cohere’s embed-english-v3 model because our text is in English, and Cohere offers free tokens for testing, making it easy to try out.
```

Comment:
```suggestion
For this tutorial, I chose to use Cohere's `embed-english-v3` model because our text is in English, and Cohere offers free tokens for testing, making it easy to try out.
```